### PR TITLE
dev/Core#53 Add in stats to AB tests and add in clickthrough and open…

### PR DESCRIPTION
**Pull request by seamuslee001**
…ed % stats to mailing reports

Overview
----------------------------------------
This adds Mailing delivery stat percentages to AB tests and also adds in new opened and clickthrough rate stats

Before
----------------------------------------
No AB testing % stats
screenshot 2018-04-09 14 20 48

After
----------------------------------------
AB Tests have % stats and rates. 

screenshot 2018-04-09 14 23 05

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@seamuslee001 will trade for #11954

**Comment by seamuslee001** [comment]
Deal @eileenmcnaughton

File: CRM/Mailing/BAO/Mailing.php [hunk 1]
```diff
@@ -1997,12 +1997,16 @@ public static function &report($id, $skipDetails = FALSE, $isSMS = FALSE) {
         $row['bounce_rate'] = (100.0 * $mailing->bounce) / $mailing->queue;
         $row['unsubscribe_rate'] = (100.0 * $row['unsubscribe']) / $mailing->queue;
         $row['optout_rate'] = (100.0 * $row['optout']) / $mailing->queue;
+        $row['opened_rate'] = ($row['opened'] / $mailing->delivered * 100.0);
```
Comment (eileenmcnaughton) [inline CRM/Mailing/BAO/Mailing.php]: We should probably add protection against a hard error if delivered === 0 (presumably a very obscure situation but worth handling I think)

Comment (seamuslee001) [reply CRM/Mailing/BAO/Mailing.php]: yeh i guess so

Comment (eileenmcnaughton) [reply CRM/Mailing/BAO/Mailing.php]: Also - can we stdise bracketting - I prefer but mostly just thing same is better 
```
 $row['opened_rate'] = ($row['opened'] / $mailing->delivered) * 100;
```
with conditional

 $row['opened_rate'] = $mailing->delivered ?  (($row['opened'] / $mailing->delivered) * 100) : '';

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I agree with this change in principle - ie it adds additional useful information to the A-B stats report and the change is localised to that report & hence low risk. I don't believe it could have a negative impact if it works as intended.

Per comments I think each divided by line has a risk of being divided by 0 so worth going through those - in fact my test did hit that on the second mailing (I added the screen shots to the PR description above - but perhaps the after might get replaced again before merge).

Visually I think

Tracked Opens 1 (100%) is better than Tracked Opens 1 100%

I think I'd also prefer the link to be on the number & not the % but from a quick look at the code that looks like more hassle than it's worth to do that.

**Comment by seamuslee001** [comment]
@eileenmcnaughton thanks for that I think i have now addressed all issues you alluded to in your review. I haven't quite got the () around on AB tests but there is at least now a ui difference between the number and the %

File: api/v3/Mailing.php [hunk 2]
```diff
@@ -769,6 +769,9 @@ function civicrm_api3_mailing_stats($params) {
         break;
     }
   }
+  $stats[$params['mailing_id']]['delivered_rate'] = round((100.0 * $stats[$params['mailing_id']]['Delivered']) / CRM_Mailing_Event_BAO_Queue::getTotalCount($params['mailing_id'], $params['job_id']), 2) . '%';
```
Comment (eileenmcnaughton) [inline api/v3/Mailing.php]: could these be subject to division by zero?

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I'm still seeing my Granny on the B side
screenshot 2018-04-09 16 11 34

**Comment by eileenmcnaughton** [comment]
Yes - there is an inconsistency in 0 vs na but that is not new in this PR. The grey power is gone!

screenshot 2018-04-09 16 25 48

**Comment by eileenmcnaughton** [comment]
I'm happy this is ready to merge now - the things I observed have been addressed

**Comment by seamuslee001** [comment]
Merging as it indicates merge-on-pass

**Pull request merged**